pymc3/distributions/dist_math.py

```diff
@@ -365,34 +365,31 @@ def conjugate_solve_triangular(outer, inner):
             grad = tt.triu(s + s.T) - tt.diag(tt.diagonal(s))
         return [tt.switch(ok, grad, floatX(np.nan))]
 
+
 class SplineWrapper (theano.Op):
     """
     Creates a theano operation from scipy.interpolate.UnivariateSpline
     """
 
-    __props__ = ('spline',)
+    __props__ = ('spline', 'grad_op')
     itypes = [tt.dscalar]
     otypes = [tt.dscalar]
 
-    def __init__(self, spline):
+    def __init__(self, spline, n_derivatives):
         self.spline = spline
 
+        if n_derivatives > 0:
+            self.grad_op = SplineWrapper(spline.derivative(), n_derivatives - 1)
+        else:
+            self.grad_op = None
+
     def perform(self, node, inputs, output_storage):
         x, = inputs
         output_storage[0][0] = np.asarray(self.spline(x))
 
-class DifferentiableSplineWrapper (SplineWrapper):
-    """
-    Creates a theano operation with defined gradient from
-    scipy.interpolate.UnivariateSpline
-    """
-
-    def __init__(self, spline):
-        super(DifferentiableSplineWrapper, self).__init__(spline)
-        self.spline_grad = SplineWrapper(spline.derivative())
-        self.__props__ += ('spline_grad',)
-
     def grad(self, inputs, grads):
+        if self.grad_op is None:
+            raise NotImplementedError
         x, = inputs
         x_grad, = grads
-        return [x_grad * self.spline_grad(x)]
+        return [x_grad * self.grad_op(x)]
```

Comment:
To avoid O(n) calculations at each call of `grad` the SciPy spline (created by `self.spline.derivative()`) should be pre-calculated and stored inside of the object. But in this case it is simpler to also store a Theano operation for gradient, rather than somehow maintain a list of pre-created SciPy splines and then pass them to constructor of a Theano operation in the grad method, so that the resulting operations also uses pre-calculated SciPy spline for the n-th order derivative.